src/Orleans/Threading/CancellationTokenHolderGrain.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Placement;
+
+namespace Orleans.Threading
+{
+    [PreferLocalPlacement]
+    internal class CancellationTokenHolderGrain : Grain, ICancellationTokenHolderGrain
+    {
+        private CancellationTokenSource cancellationTokenSource;
+        private bool collectionScheduled;
+
+        public override Task OnActivateAsync()
+        {
+            DelayDeactivation(TimeSpan.FromDays(5));
+            cancellationTokenSource = new CancellationTokenSource();
+            return TaskDone.Done;
+        }
+
+        public Task<CancellationToken> GetCancellationToken()
+        {
+            return Task.FromResult(cancellationTokenSource.Token);
+        }
+
+        public Task Cancel()
+        {
+            cancellationTokenSource.Cancel();
+            DelayDeactivation(TimeSpan.FromSeconds(30));
```

Comment:
Sorry, thought about completly different number. About this one: in case of cansellation happened before token requesting this will be unreliable with 0.1 ms. 30 seconds should be sufficient for almost all cases, as default request timeout is 30 seconds too.